### ISSUE TITLE
Trying to build ASP.NET Core with 1.0.0-dev local build

### ISSUE THREAD

**Issue by DRSDavidSoft** [peachpiecompiler/peachpie#867]
I hope this is not something that I overlooked, but I'm getting a **The build failed. Fix the build errors and run again.** when trying to build the [web sample](https://github.com/iolevel/peachpie-samples/tree/master/web-application) using a local build of Peach Pie. The same thing works in cli!

The process:
- `git clone`d the repository before
- `git pull`ed from master
- `dotnet clean`
- `dotnet restore`
- invoke `build/build.ps1` from Powershell  
    (here's the only modification I made to the script, since `msbuild.exe` is not in my path but `dotnet.exe` is)  
    ```patch
    --- a/build/build.ps1
    +++ b/build/build.ps1
    @@ -7,6 +7,6 @@ $props = "/p:Version=$version,Configuration=$config"
    
     Write-Host -f Green "msbuild $props ..."
    
    -msbuild /t:restore "$PSScriptRoot/../Peachpie.sln" /v:m $props
    +dotnet msbuild /t:restore "$PSScriptRoot/../Peachpie.sln" /v:m $props
     dotnet restore "$PSScriptRoot/../src/Peachpie.NET.Sdk" /v:m $props
    -msbuild /t:build "$PSScriptRoot/../Peachpie.sln" /v:m $props
    +dotnet msbuild /t:build "$PSScriptRoot/../Peachpie.sln" /v:m $props
    ```
- (alternatively, I also tried `dotnet build` in the root path)
- `dotnet restore`, not sure if it's required
- **`build/update-cache.ps1`**

Then, on the `web-application` directory:
- update any references to `1.0.0-dev` (e.g. in `*.csproj`, `*.msbuildproj`)
- `dotnet clean`
- `dotnet restore`
- `dotnet run -p app`

Here's the exception and its stack trace:
```cs
Process terminated. Assertion Failed
   at Pchp.CodeAnalysis.Semantics.VariableReferenceExtensions.EmitStore(IPlace place, CodeGenerator cg, LhsStack& lhs, TypeSymbol stack, BoundAccess access)
   at Pchp.CodeAnalysis.Semantics.LocalVariableReference.EmitStore(CodeGenerator cg, LhsStack& lhs, TypeSymbol stack, BoundAccess access)
   at Pchp.CodeAnalysis.Semantics.BoundAssignEx.Emit(CodeGenerator cg)
   at Pchp.CodeAnalysis.Semantics.BoundExpressionStatement.Emit(CodeGenerator cg)
   at Pchp.CodeAnalysis.Semantics.BoundStatement.Pchp.CodeAnalysis.CodeGen.IGenerator.Generate(CodeGenerator cg)
   at Pchp.CodeAnalysis.CodeGen.CodeGenerator.Generate(IGenerator element)
   at System.Collections.Generic.List`1.ForEach(Action`1 action)
   at Pchp.CodeAnalysis.Semantics.Graph.BoundBlock.Emit(CodeGenerator cg)
   at Pchp.CodeAnalysis.Semantics.Graph.StartBlock.Emit(CodeGenerator cg)
   at Pchp.CodeAnalysis.Semantics.Graph.BoundBlock.Pchp.CodeAnalysis.CodeGen.IGenerator.Generate(CodeGenerator cg)
   at Pchp.CodeAnalysis.CodeGen.CodeGenerator.Generate(IGenerator element)
   at Pchp.CodeAnalysis.CodeGen.CodeGenerator.GenerateBlock(BoundBlock block)
   at Pchp.CodeAnalysis.CodeGen.CodeGenerator.LocalScope.ContinueWith(BoundBlock block)
   at Pchp.CodeAnalysis.CodeGen.CodeGenerator.GenerateScope(BoundBlock block, ScopeType type, Int32 to)
   at Pchp.CodeAnalysis.CodeGen.CodeGenerator.GenerateScope(BoundBlock block, Int32 to)
   at Pchp.CodeAnalysis.Symbols.SourceRoutineSymbol.Generate(CodeGenerator cg)
   at Pchp.CodeAnalysis.CodeGen.CodeGenerator.Generate()
   at Pchp.CodeAnalysis.CodeGen.MethodGenerator.<>c__DisplayClass1_0.<GenerateMethodBody>b__0(ILBuilder builder)
   at Pchp.CodeAnalysis.CodeGen.MethodGenerator.GenerateMethodBody(PEModuleBuilder moduleBuilder, MethodSymbol routine, Action`1 builder, VariableSlotAllocator variableSlotAllocatorOpt, DiagnosticBag diagnostics, Boolean emittingPdb)
   at Pchp.CodeAnalysis.CodeGen.MethodGenerator.GenerateMethodBody(PEModuleBuilder moduleBuilder, SourceRoutineSymbol routine, Int32 methodOrdinal, VariableSlotAllocator variableSlotAllocatorOpt, DiagnosticBag diagnostics, Boolean emittingPdb)
   at Pchp.CodeAnalysis.SourceCompiler.EmitMethodBody(SourceRoutineSymbol routine)
   at System.Threading.Tasks.Parallel.<>c__DisplayClass44_0`2.<PartitionerForEachWorker>b__1(IEnumerator& partitionState, Int32 timeout, Boolean& replicationDelegateYieldedBeforeCompletion)
   at System.Threading.Tasks.TaskReplicator.Replica`1.ExecuteAction(Boolean& yieldedBeforeCompletion)
   at System.Threading.Tasks.TaskReplicator.Replica.Execute()
   at System.Threading.Tasks.TaskReplicator.Replica.<>c.<Execute>b__7_0(Object s)
   at System.Threading.Tasks.Task.InnerInvoke()
   at System.Threading.Tasks.Task.<>c.<.cctor>b__274_0(Object obj)
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
   at System.Threading.Tasks.Task.ExecuteEntryUnsafe(Thread threadPoolThread)
   at System.Threading.Tasks.Task.ExecuteFromThreadPool(Thread threadPoolThread)
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback()


The build failed. Fix the build errors and run again.
```

I'm not sure what assertion is being failed.  The same workflow works fine when building cli projects.

Here's a test cli script:
```php
echo PHP_VERSION . PHP_EOL . PEACHPIE_VERSION . PHP_EOL;
```

This results in:
```
7.4.69-peachpie
1.0.0-dev+dd1354fc4084dc08559fa067206ee526ec64985d
```

What can I try in order to resolve this issue?

**Comment by jakubmisek** [comment]
it's a debug assertion, used for peachpie development itself;

Anyways it shows that in the PHP code you're compiling is something wrong (or unexpected).

if you want to dig in, you can try running
```shell
cd app
dotnet build -p:DebuggerAttach=true
```

you can then debug the assertion - check the `cg` variable - `cg.Routine.Name` and `cg.Routine.ContainingFile.RelativeFilePath` will give you a hint about the function causing this assertion :)

**Comment by DRSDavidSoft** [comment]
This is awesome, I wish I had a deeper understanding of PeachPie!  Thank you for the tip, I followed the CodeGenerator properties and the `cg.Routine.Name` pointed to the piece of code in my project that I think is causing this.  Here's the assertion in `EmitSore`:

```cs
            if (access.IsWriteRef)
            {
                Debug.Assert(stack == cg.CoreTypes.PhpAlias);
                //           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Here's the relevant user code that is in the mentioned `cg.Routine.Name`:
```php
public function fooBar() {
            // …
            $_ = &$this;
```

There is a good reason to do this BTW, since I'm referencing it from a closure (anonymous function) later in a variable that this method returns:

```php
            // …
            $actions['foo'] = function ( $arg1 ) use (&$_) { … }

            return $actions;
}
```

So, I have a couple of questions now:
1. Why is this assertion being thrown? Is it expected?
2. Why is it _only_ thrown when using the `1.0.0-dev` reference, and not the one from AppVeyor?
3. Since this is valid PHP code and works (unless I'm mistaken, in which case please inform me) -- is there a way to resolve this issue in PeachPie?

Thanks! :D

**Comment by jakubmisek** [comment]
nice!

1. It's thrown because it is an unexpected/unhandled situation; that's why the Debug assertion is there. 

2. Because your `1.0.0-dev` is built using debug configuration. AppVeyor is then built using Release configuration.

3. It is a valid PHP code, but still unexpected by the compiler; in .NET we are not able to make a reference to `$this` variable because `$this` variable cannot be modified (even in PHP), so making a reference to `$this` does not make sense even in PHP.

following code for example in PHP:
```php
class X {
	function f() {
		$_ = &$this;
		$_ = null;
	}
}
```
leaves $this unmodified, and does not even report a warning in PHP (even tho `$this = null;` would result in fatal error)

So when PeachPie is built using Release configuration, we silently ignore it as well, but the debug assertion reminds us, there is something wrong, and we should probably report a warning in compile time maybe?

**Comment by jakubmisek** [comment]
@DRSDavidSoft the assertion should be fixed now, as it behaves like PHP, we'll add a test case for it as well

**Comment by DRSDavidSoft** [comment]
@jakubmisek Thank you for the apprehensive answers!  I understand now.  Also, thank you for the fix!

One small question though, is `$_ = &$this;` unnecessary in this context?  I mean we never intend to assign anything _to_ `$this`, but `$_` should _always_ be referenced to it.  I understand that `$_ = null;` → `$this = null;`. which is invalid, but if `&` is omitted, will `$_` still be a reference to it?

I enjoy that PeachPie can analyze this kind of issues and warn the user, warning the user of valid PHP code that doesn't really make sense, as stated in [blog post](https://www.peachpie.io/2020/06/diagnosing-the-internet.html).  Even though PHP is not strict about it, this kind of (in)valid code needs to be fixed in the user codebase, and I appreciate that PeachPie allows the user to be aware of these issues.

In summary, I'd like to thank you for your continued efforts!
